azure/services/bastionhosts/spec.go

```diff
@@ -40,69 +41,63 @@ type AzureBastionSpec struct {
-func (s *AzureBastionSpec) Parameters(ctx context.Context, existing interface{}) (parameters interface{}, err error) {
-	if existing != nil {
-		if _, ok := existing.(armnetwork.BastionHost); !ok {
-			return nil, errors.Errorf("%T is not an armnetwork.BastionHost", existing)
-		}
-		// bastion host already exists
-		return nil, nil
+// Parameters implements azure.ASOResourceSpecGetter.
+func (s *AzureBastionSpec) Parameters(ctx context.Context, existingBastionHost *asonetworkv1.BastionHost) (parameters *asonetworkv1.BastionHost, err error) {
+	bastionHost := &asonetworkv1.BastionHost{}
+	bastionHost.Spec = asonetworkv1.BastionHost_Spec{}
+	if existingBastionHost != nil {
+		bastionHost = existingBastionHost
 	}
 
 	bastionHostIPConfigName := fmt.Sprintf("%s-%s", s.Name, "bastionIP")
-
-	return armnetwork.BastionHost{
-		Name:     ptr.To(s.Name),
-		Location: ptr.To(s.Location),
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-			ClusterName: s.ClusterName,
-			Lifecycle:   infrav1.ResourceLifecycleOwned,
-			Name:        ptr.To(s.Name),
-			Role:        ptr.To("Bastion"),
-		})),
-		SKU: &armnetwork.SKU{
-			Name: ptr.To(armnetwork.BastionHostSKUName(s.Sku)),
-		},
-		Properties: &armnetwork.BastionHostPropertiesFormat{
-			EnableTunneling: ptr.To(s.EnableTunneling),
-			DNSName:         ptr.To(fmt.Sprintf("%s-bastion", strings.ToLower(s.Name))),
-			IPConfigurations: []*armnetwork.BastionHostIPConfiguration{
-				{
-					Name: ptr.To(bastionHostIPConfigName),
-					Properties: &armnetwork.BastionHostIPConfigurationPropertiesFormat{
-						Subnet: &armnetwork.SubResource{
-							ID: &s.SubnetID,
-						},
-						PublicIPAddress: &armnetwork.SubResource{
-							ID: &s.PublicIPID,
-						},
-						PrivateIPAllocationMethod: ptr.To(armnetwork.IPAllocationMethodDynamic),
-					},
+	bastionHost.Spec.AzureName = s.Name
+	bastionHost.Spec.Location = ptr.To(s.Location)
+	bastionHost.Spec.Owner = &genruntime.KnownResourceReference{
+		Name: s.ClusterName,
+	}
+	bastionHost.Spec.Tags = infrav1.Build(infrav1.BuildParams{
+		ClusterName: s.ClusterName,
+		Lifecycle:   infrav1.ResourceLifecycleOwned,
+		Name:        ptr.To(s.Name),
+		Role:        ptr.To("Bastion"),
+	})
+	bastionHost.Spec.Sku = &asonetworkv1.Sku{
+		Name: ptr.To(asonetworkv1.Sku_Name(s.Sku)),
+	}
+	bastionHost.Spec.EnableTunneling = ptr.To(s.EnableTunneling)
+	bastionHost.Spec.DnsName = ptr.To(fmt.Sprintf("%s-bastion", strings.ToLower(s.Name)))
+	bastionHost.Spec.IpConfigurations = []asonetworkv1.BastionHostIPConfiguration{
+		{
+			Name: ptr.To(bastionHostIPConfigName),
+			Subnet: &asonetworkv1.BastionHostSubResource{
+				Reference: &genruntime.ResourceReference{
+					ARMID: s.SubnetID,
+				},
+			},
+			PublicIPAddress: &asonetworkv1.BastionHostSubResource{
+				Reference: &genruntime.ResourceReference{
+					ARMID: s.PublicIPID,
 				},
 			},
+			PrivateIPAllocationMethod: ptr.To(asonetworkv1.IPAllocationMethod_Dynamic),
```

Comment:
Is it valid and possible to have a bastionHost with IPConfigurations such that the one of IPConfiguration just has `ptr.To(asonetworkv1.IPAllocationMethod_Dynamic)`?
I mean is the below configuration valid and possible?
```yaml
spec:
	.
	.
	ipConfigurations:
		- publicIPAddress: "Dynamic"
	.
```

Reply:
The reason I ask is because the [spec_test/get result as BastionHost with empty values when existing BastionHost is present with empty data](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4143/files#diff-0894c31cc81cb639d7b041d520ea154ffd418c6605a41cd8c444e17cc5d78b1cR102-R127) appears to be returning a spec which is not completely nil (it has a `ipConfigurations` with one element in it). I am wondering if it should be `nil`.

Reply:
It's returning a spec that has the Subnet and PublicIPAddress specs populated. The only difference is that the ARMID field will have an empty string. I'm not sure if that's what we want but this goes back to not returning existingBastionHost early if it is not nil https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4143#discussion_r1372234691. Any extra fields that aren't specified will have weird outputs such as this https://github.com/willie-yao/cluster-api-provider-azure/blob/aso-bastionhosts/azure/services/bastionhosts/spec_test.go#L80-L81